crates/sui-core/src/authority_active/cancellable_spawner.rs

```diff
@@ -0,0 +1,120 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::future::Future;
+use std::sync::Arc;
+use tokio::sync::{broadcast, OwnedRwLockWriteGuard, RwLock};
+use tokio::task::JoinHandle;
+use tracing::{debug, trace};
+
+use tap::TapFallible;
+
+use sui_types::error::{SuiError, SuiResult};
+
+/// CancellableSpawner allows spawned tasks to be cancelled easily without keeping track of every
+/// JoinHandle that gets created.
+pub struct CancellableSpawner {
+    cancel_tx: broadcast::Sender<()>,
+    spawn_lock: Arc<RwLock<()>>,
+    task_lock: Arc<RwLock<()>>,
+}
+
+/// While a CancelGuard is held, no new tasks can be spawned on the CancellableSpawner.
+pub struct CancelGuard(OwnedRwLockWriteGuard<()>);
+
+impl CancellableSpawner {
+    pub fn new() -> Self {
+        let (cancel_tx, _) = broadcast::channel(1);
+        Self {
+            cancel_tx,
+            spawn_lock: Default::default(),
+            task_lock: Default::default(),
+        }
+    }
+
+    /// Spawn a future via tokio::spawn - may fail if cancel_all_tasks has been called.
+    pub fn spawn<F>(&self, fut: F) -> SuiResult<JoinHandle<()>>
+    where
+        F: Future<Output = ()> + Send + 'static,
+    {
+        // if spawn_lock is held for write, the spawner is in the process of cancelling, so new
+        // spawn attempts must fail.
+        let _spawn_guard = self
+            .spawn_lock
+            .try_read()
+            .tap_err(|_| {
+                debug!("task could not be spawned, CancellableSpawner::cancel() is in progress")
+            })
+            .map_err(|_| SuiError::TaskSpawnError)?;
+
+        // While spawn_lock is held, task_lock.write() cannot be called, so try_read must succeed.
+        let task_guard = self.task_lock.clone().try_read_owned().unwrap();
```

Comment:
I think i would probably avoid using a lock, because i think its error prone, and instead allocate some object that can be used as a notification once a particular task has completed. I mean you already have access to the JoinHandle from the tokio::spawn call, why not track those and call abort on those explicitly? (you can then await the join handles if you really want to wait for drop logic to run)

Reply:
the reason i didn't keep the join handles around to call abort() on is that then I would have to have a container to hold the handles, and then write additional code to clean up any handles that finished on their own. (Or else let the container grow without bound).

That's not insurmountable, but this seemed like less code, simpler code, and thus less error prone.

Can you explain why this technique might cause problems? (beyond just saying that locks are hard?)